spmd/tensor/ops/view_ops.py

```diff
@@ -0,0 +1,454 @@
+    """
+    assert len(in_shard) == len(mesh_sizes)
+    # print('local_output_shape:', in_shard, local_in_shape, rule, mesh_sizes)
+
+    sharded_in_dims = set(s.dim for s in in_shard if isinstance(s, Shard))
+
+    def get_dim_size(cmd):
+        if isinstance(cmd, int):
+            return (
+                local_in_shape[cmd],
+                cmd if cmd in sharded_in_dims else None
+            )
+
+        elif cmd[0] == 'flatten':
+            for in_dim in cmd[1][1:]:
+                assert not in_dim in sharded_in_dims, 'Only the first member of a FLATTEN group can be sharded'
+            return (
+                _prod(get_dim_size(a)[0] for a in cmd[1]),
+                cmd[1][0] if cmd[1][0] in sharded_in_dims else None
+            )
+        elif cmd[0] == 'comp':
+            if cmd[2] > 0:
+                # we will shard only on the first dimension of the group
+                # so the shape should be the nominal shape of the component
+                return cmd[1][cmd[2]], None
+            else:
+                dim_size, in_dim = get_dim_size(cmd[1])
+                if in_dim is None:
+                    # in case input dim is not sharded; our size will be
+                    # the size of the corresponding input
+                    return dim_size, None
+                # we need to check that the input dimension is divisble
+                # by the size of the submesh we're sharding it on
+                submesh_size = 1
+                for size, shard in zip(mesh_sizes, in_shard):
+                    if isinstance(shard, Shard) and shard.dim == in_dim:
+                        submesh_size *= size
+                out_size = cmd[1][0]
+                assert out_size % submesh_size, 'Resulting dimension size is not divisible by its mesh dimension.'
+                return out_size // submesh_size, in_dim
+        elif cmd[0] == 'singleton':
+            return 1, None
+        elif cmd[0] == 'broadcast':
+            return cmd[2], None
+        elif cmd[0] == 'newdim':
+            return cmd[1], None
+        elif cmd[0] == 'repeat':
+            assert not cmd[1] in sharded_in_dims, 'Cannot tile sharded dimension.'
+            return local_in_shape[cmd[1]], None
+        else:
+            raise RuntimeError(f'cmd not found: {cmd}, in rule: {rule}')
+ 
+    dim_map = {}
+    out_shape = []
+    for dim, cmd in enumerate(rule):
+        out_size, in_dim = get_dim_size(cmd)
+        out_shape.append(out_size)
+        if in_dim is not None:
+            dim_map[in_dim] = dim
+
+    return (
+        out_shape,
+        [Shard(dim_map[s.dim]) if isinstance(s, Shard) else s for s in in_shard]
+    )
+
+
+def register_prop_rule_map(aten_op_name, local_op_name):
+
+    @register_prop_rule(aten_op_name)
+    def reshape_prop(op_schema: OpSchema):
+        spec = ops[local_op_name]
+
+        # note we are passing _global_ tensors
+        rules = spec.dim_map(*op_schema.args, **op_schema.kwargs)
```

Comment:
this is where i need the global shapes as well as the args for sharding propragation.